// cn-division/index.js
/**
 * cn-division - 中国行政区划数据 2026版
 * 基于民政部地名服务（dmfw）最新年版数据
 */

import pca from './dist/no-code/pca.json' with { type: 'json' };
import pc from './dist/no-code/pc.json' with { type: 'json' };
import provinces from './dist/no-code/provinces.json' with { type: 'json' };
import cities from './dist/no-code/cities.json' with { type: 'json' };
import counties from './dist/no-code/counties.json' with { type: 'json' };

// 带编码数据
import pcaCode from './dist/code/pca.json' with { type: 'json' };
import provincesCode from './dist/code/provinces.json' with { type: 'json' };
import citiesCode from './dist/code/cities.json' with { type: 'json' };
import countiesCode from './dist/code/counties.json' with { type: 'json' };

/**
 * 根据编码获取名称
 * @param {string} code - 行政区划编码
 * @returns {string | undefined}
 */
export function getNameByCode(code) {
  if (!code) return undefined;
  const codeStr = String(code);

  if (codeStr.length === 12) {
    const county = getCountyByCode(codeStr);
    if (county) return county.n;
  }

  if (codeStr.length === 2 || codeStr.endsWith('0000')) {
    const province = getProvinceByCode(codeStr);
    if (province) return province.n;
  }

  if (codeStr.length === 4 || codeStr.endsWith('00')) {
    const city = getCityByCode(codeStr);
    if (city) return city.n;
  }

  const county = getCountyByCode(codeStr);
  if (county) return county.n;

  return undefined;
}

// 导出数据
export { pca, pc, provinces, cities, counties };
// 导出带编码数据
export { pcaCode, provincesCode, citiesCode, countiesCode };

// 不带编码列表数据
import provincesNoCode from './dist/no-code/provinces.json' with { type: 'json' };
import citiesNoCode from './dist/no-code/cities.json' with { type: 'json' };
import countiesNoCode from './dist/no-code/counties.json' with { type: 'json' };
export { provincesNoCode, citiesNoCode, countiesNoCode };

// 常量
export const MUNICIPALITY_CODES = ['11', '12', '31', '50'];
export const VERSION = '2026.0.0';

// 内部缓存（lazy init，转换只发生一次）
let _cacheDataWithCode = null;
let _cacheData = null;
let _cachePCWithCode = null;

/**
 * 获取省市区三级 Cascader 数据（带编码版）
 * value = 行政区划编码（string），label = 名称
 * 兼容 antd Cascader 和 TDesign t-cascader，无需 fieldNames/keys 配置
 */
export function getCascaderDataWithCode() {
  return _cacheDataWithCode || (_cacheDataWithCode = pcaCode.map(province => ({
    label: province.n,
    value: String(province.c),
    children: province.ch && province.ch.map(city => ({
      label: city.n,
      value: String(city.c),
      children: city.ch && city.ch.map(county => ({
        label: county.n,
        value: String(county.c),
      }))
    }))
  })));
}

/**
 * 获取省市区三级 Cascader 数据（不带编码版）
 * value = 名称（string）
 */
export function getCascaderData() {
  return _cacheData || (_cacheData = Object.entries(pca).map(([province, cities]) => ({
    label: province,
    value: province,
    children: Object.entries(cities).map(([city, counties]) => ({
      label: city,
      value: city,
      children: counties.map(county => ({ label: county, value: county }))
    }))
  })));
}

/**
 * 获取省市二级 Cascader 数据（带编码版）
 * value = 行政区划编码（string），label = 名称
 * 适合查价格等只需省市的场景
 */
export function getCascaderPCWithCode() {
  return _cachePCWithCode || (_cachePCWithCode = pcaCode.map(province => ({
    label: province.n,
    value: String(province.c),
    children: province.ch && province.ch.map(city => ({
      label: city.n,
      value: String(city.c),
    }))
  })));
}

/**
 * 根据省级代码获取省份信息
 */
export function getProvinceByCode(code) {
  if (!code) return undefined;
  const provinceCode = String(code).substring(0, 2);
  return provincesCode.find(p => String(p.c).padStart(2, '0') === provinceCode);
}

/**
 * 根据市级代码获取城市信息
 */
export function getCityByCode(code) {
  if (!code) return undefined;
  const cityCode = String(code).substring(0, 4);
  return citiesCode.find(c => String(c.c) === cityCode);
}

/**
 * 根据区县代码获取区县信息
 */
export function getCountyByCode(code) {
  if (!code) return undefined;
  const countyCode = String(code);
  return countiesCode.find(a => String(a.c) === countyCode);
}

/**
 * 根据三级行政区划编码构建统一 AddressValue 对象
 * @param {string|number} provinceCode
 * @param {string|number} cityCode
 * @param {string|number} countyCode
 * @returns {AddressValue|null}
 */
export function getAddressValueFromCodes(provinceCode, cityCode, countyCode) {
  const pStr = String(provinceCode);
  const ctStr = String(cityCode);
  const coStr = String(countyCode);
  const prov = getProvinceByCode(pStr);
  const city = getCityByCode(ctStr);
  const county = getCountyByCode(coStr);
  if (!prov) return null;
  return {
    province: { code: pStr, name: prov.n },
    city: { code: ctStr, name: city ? city.n : '' },
    county: { code: coStr, name: county ? county.n : '' }
  };
}

/**
 * 检查是否为直辖市代码
 */
export function isMunicipalityCode(code) {
  if (!code) return false;
  const provinceCode = String(code).substring(0, 2);
  return MUNICIPALITY_CODES.includes(provinceCode);
}

/**
 * 获取省份下的所有城市
 */
export function getCitiesByProvince(provinceCode) {
  if (!provinceCode) return [];
  const targetProvinceCode = String(provinceCode).substring(0, 2);
  return citiesCode.filter(c => c.p === targetProvinceCode);
}

/**
 * 获取城市下的所有区县
 */
export function getCountiesByCity(cityCode) {
  if (!cityCode) return [];
  const targetCityCode = String(cityCode).substring(0, 4);
  return countiesCode.filter(a => a.cc === targetCityCode);
}

/**
 * 验证行政区划代码格式（6位或12位数字）
 */
export function validateCode(code) {
  if (!code || typeof code !== 'string') return false;
  if (!/^(\d{6}|\d{12})$/.test(code)) return false;
  return code.substring(0, 2) !== '00';
}

/**
 * 格式化显示完整的行政区划路径
 */
export function formatFullPath(code) {
  if (!validateCode(code)) return '';
  const parts = [];
  const province = getProvinceByCode(code);
  if (province) parts.push(province.n);
  const city = getCityByCode(code);
  if (city && (!province || city.n !== province.n)) parts.push(city.n);
  const county = getCountyByCode(code);
  if (county) parts.push(county.n);
  return parts.join(' / ');
}

/**
 * 简单的地址解析函数
 * 从地址字符串中提取省市区信息
 */
export function parseAddress(address) {
  if (!address || typeof address !== 'string') return null;
  const result = { province: undefined, city: undefined, county: undefined, rest: address };
  let remaining = address.trim();
  for (const p of provincesNoCode) { if (remaining.includes(p)) { result.province = p; remaining = remaining.replace(p, ''); break; } }
  for (const c of citiesNoCode) { if (remaining.includes(c)) { result.city = c; remaining = remaining.replace(c, ''); break; } }
  for (const a of countiesNoCode) { if (remaining.includes(a)) { result.county = a; remaining = remaining.replace(a, ''); break; } }
  result.rest = remaining.trim();
  if (!result.province && !result.city && !result.county) return null;
  return result;
}

// 默认导出
export default {
  pca, pc,
  provinces, cities, counties,
  provincesNoCode, citiesNoCode, countiesNoCode,
  pcaCode, provincesCode, citiesCode, countiesCode,
  getNameByCode, getCascaderDataWithCode, getCascaderData, getCascaderPCWithCode,
  getAddressValueFromCodes, getProvinceByCode, getCityByCode, getCountyByCode,
  parseAddress, isMunicipalityCode, getCitiesByProvince, getCountiesByCity,
  validateCode, formatFullPath,
  MUNICIPALITY_CODES, VERSION,
};
